packages/devtools/lib/src/timeline/cpu_flame_chart.dart

```diff
@@ -0,0 +1,169 @@
+    return color;
+  }
+
+  @override
+  void reset() {
+    super.reset();
+    _colorOffset = 0;
+  }
+
+  // TODO(kenzie): rewrite this render method to draw to canvas.
+  @override
+  void render() {
+    final CpuProfileData cpuProfileData = data;
+    final totalWidth = element.clientWidth - 2 * flameChartInset;
+
+    final Map<String, double> stackFrameLefts = {};
+
+    double calculateLeftForStackFrame(CpuStackFrame stackFrame) {
+      double left;
+      if (stackFrame.parent == null) {
+        left = flameChartInset.toDouble();
+      } else {
+        final stackFrameIndex = stackFrame.parent.children.indexOf(stackFrame);
+        if (stackFrameIndex == 0) {
+          // This is the first child of parent. [left] should equal the left
+          // value of [stackFrame]'s parent.
+          left = stackFrameLefts[stackFrame.parent.id];
+        } else {
+          // [stackFrame] is not the first child of its parent. [left] should
+          // equal the right value of its previous sibling.
+          final previous = stackFrame.parent.children[stackFrameIndex - 1];
+          left = stackFrameLefts[previous.id] +
+              (totalWidth * previous.cpuConsumptionRatio);
+        }
+      }
+      stackFrameLefts[stackFrame.id] = left;
+      return left;
+    }
+
+    void drawRecursively(CpuStackFrame stackFrame, int row) {
+      final double width =
+          totalWidth * stackFrame.cpuConsumptionRatio - samplePadding;
+
+      final item = CpuFlameChartItem(
+        stackFrame,
+        calculateLeftForStackFrame(stackFrame),
+        width,
+        row * FlameChart.rowHeight + FlameChart.padding,
+        nextColor(),
+        Colors.black,
+        Colors.black,
+      );
+      addItemToFlameChart(item, this);
+
+      for (CpuStackFrame child in stackFrame.children) {
+        drawRecursively(
+          child,
+          row + 1,
+        );
+      }
+    }
+
+    drawRecursively(cpuProfileData.cpuProfileRoot, 0);
+  }
+
+  @override
+  void updateChartForZoom() {
+    super.updateChartForZoom();
+    element.scrollLeft = math.max(0, floatingPointScrollLeft.round());
+  }
+}
+
+class CpuFlameChartItem extends FlameChartItem {
+  CpuFlameChartItem(
+    this.stackFrame,
+    num startingLeft,
+    num startingWidth,
+    num top,
+    Color backgroundColor,
+    Color defaultTextColor,
+    Color selectedTextColor,
+  ) : super(
+          startingLeft: startingLeft,
+          startingWidth: startingWidth,
+          top: top,
+          backgroundColor: backgroundColor,
+          defaultTextColor: defaultTextColor,
+          selectedTextColor: selectedTextColor,
+        );
+
+  final CpuStackFrame stackFrame;
+
+  @override
+  void setText() {
+    element.title = stackFrame.toString();
+
+    if (stackFrame.parent == null) {
+      itemLabel.text = stackFrame.toString();
+    } else {
+      itemLabel.text = stackFrame.name;
+    }
+  }
+
+  @override
+  void setOnClick() {
+    element.onClick.listen((e) {
```

Comment:
fyi one optimization is we are adding a top of event listeners. You can add one global event listener for the overall view.

Reply:
Are you suggesting I add a global listener and then figure out which element was clicked based on mouse position?

Reply:
added a todo.